### ISSUE TITLE
Explicit batch_size when initializing the model

### ISSUE THREAD

**Issue by thereisnospooon** [auvisusAI/detr-tensorflow#5]
Hello, 
Thanks for the awsome implementation!

I'm trying to use the model but having trouble with the explicit definition of batch_size.

Currently when initializing a DETR model we need to specify batch size. From what I understand
every time we want to call our model on some input it needs to be in this batch_size sizing. This is OK when training but 
is it possible to change the batch size for inference?

**Comment by SchernHe** [comment]
Hey,

I see the point why this might be troubling and I just created a PR to extract the data feeding from the DETR initialization. 
Take a look [here](https://github.com/auvisusAI/detr-tensorflow/pull/6) and please comment in case of better proposals or remarks. 

Moreover, I included a notebook where you can check that you can do inference for any batch size, independent of your training (check out the first dimension of `positional_encodings` and `image`).

**Comment by thereisnospooon** [comment]
Hey,
I looked at the notebook and I think that there is maybe a bug there? The batch_size is not specified when initializing DETR...

On the other hand, I think that I have a way to solve this (please let me know if I got something wrong...) - 
1. We don't need to specify the batch_size when initializing the positional encoding. That is, keep the function create_positional_encoding the same just don't tf.repeat it to be of size batch_size and don't tf.reshape it. This will work because of broadcasting.
2. The other place when there is use of the batch_size is in the MultiHeadAttention layer. But there we get the batch_size from tf.shape(q)[0]. Even though this is not the batch_size that we specify in DETR __init__ since q is sometimes dependent on the batch_size we need to initialize the "tgt" variable in transformer.py like this - 
tgt = tf.zeros(tf.concat([tf.shape(inp)[0], tf.shape(query_pos)], axis=0))

This way "tgt" gets None to be its batch size (first dimension) and we can use any size we want in training and inference.

Hope this would help and that I explained it well.

**Comment by SchernHe** [comment]
Hey,

sorry for the late response. Little bit busy at the moment :). 

So its not working for you or why do you think its a bug? The goal of the PR was that the model is unaware of the explicit `batch_size` during `__init__` and `build_model()`, but is able to handle whatever comes from the `data_feeder`. 


**Regarding the batch_size in `DETR.build_model()`:**
- Thats not the actual batch_size we use in training, but only the placeholder for size of batches is coming from `transformer_input`. 
- However, I think it would be nice to find another tf method that just combines axes 1 and 2, such that we do not have to use tf.reshape here

**Regarding the batch_size in `MultiHeadAttention `:**
- Should be the same here. Whatever comes in from q ( which we know is of shape batch_size, seq_len, dim_transformer), we exract the first dimension (batch_size) to use in `split_heads`. 

Unfortunately, I do not yet 100% understand the two implementations. For what I understand, the major difference is that we specify the the first dimension in `tgt` (None) right now in `DETR.build_model()` via `query_pos` and you would like to move this into the `Transformer()`? 

Please correct me, if I misunderstood anything.

**Comment by java-abhinav07** [comment]
I think @thereisnospooon is absolutely correct. Things work fine during training however, due to ```batch_size = tf.shape(q)[0]``` things break down during inference. This bug remains in the notebook. In my situation, upon loading the saved model I get this bug:

![Screenshot 2020-09-20 14:22:26](https://user-images.githubusercontent.com/42839106/93707501-c934fa00-fb4c-11ea-8867-6ea33db60d6c.png)

Turns out due the ```NoneType``` tensorshape from ``q```, ```load_model``` breaks when dense layer is called. This error vanishes if I use a hard coded ```batch_size```.
